File: Solutions/python/2022/day_06.py
import sys, os
sys.path.append(os.path.abspath(os.path.join(os.path.realpath(os.path.dirname(__file__)), "..", "SetupSolutions")))

from SetupSolutions import getPuzzleInput

year = os.path.basename(os.path.dirname(__file__))
day = os.path.basename(__file__).split(".")[0].capitalize()
puzzleInput = [line.replace("\n", "") for line in getPuzzleInput(f"{year}", f"{day}.txt")]
puzzle = puzzleInput[0]

def allDifferentChars(s: str):
    existing_chars = set()
    for char in s:
        if char in existing_chars:
            return False
        else:
            existing_chars.add(char)
    
    return True

def getStartOfMarker(puzzle, interval):

    for i in range(0, len(puzzle)):
        latest = i + interval
        snip = puzzle[i:latest]

        if allDifferentChars(snip):
            return latest


# Script runs here
print(f"Advent of Code - {year}: {day.replace('_', ' ')}")
print("Part-One Answer:", getStartOfMarker(puzzle, 4))
print("Part-Two Answer:", getStartOfMarker(puzzle, 14))

File: Solutions/javascript/2022/day_06.js
const { getPuzzleInput } = require("../SetupSolutions/SetupSolutions.js");

const year = "2022";
const day = "06";
const puzzleInput = getPuzzleInput(year, `Day_${day}.txt`);
const puzzle = puzzleInput[0];

const allDifferentChars = (s) => {
    const existingChars = {};
    for (c of s.split("")) {
        if (c in existingChars) return false;
        else existingChars[c] = c;
    }
    
    return true;
}

const getStartOfMarker = (puzzle, interval) => {
    for (let i = 0; i <= puzzle.length; i++) {
        let latest = i + interval;
        let snip = puzzle.slice(i, latest)
        if (allDifferentChars(snip)) return latest;
    }
}

// Script runs here
console.log(`Advent of Code - ${year}: Day ${day}`);
console.log("Part-One Answer:", getStartOfMarker(puzzle, 4));
console.log("Part-Two Answer:", getStartOfMarker(puzzle, 14));